### ISSUE TITLE
Ability to blacklist header contents in access logs

### ISSUE THREAD

**Issue by indigoid** [containous/traefik#2067]
### Feature request

I would like to be able to configure Traefik to not log the contents of named headers. Mostly `Authorization`, but probably others in the future.

This would help us avoid exposing and archiving header content that should be considered authentication secrets.

Related to #711.

### What did you do?

Running `experimental`-tag Traefik on Kubernetes with a mostly-defaults configuration, but with JSON logging to stdout. These logs are ingested by Fluentd and then fed to Graylog for alerting/analysis. This all works very well and has been quite reliable.

### What did you expect to see?

Header values logged to Graylog along with requests, but without contents of `Authorization` and other well-known headers that are used to convey secrets.

### What did you see instead?

All header values logged to Graylog along with requests.

### Output of `traefik version`: (_What version of Traefik are you using?_)

We are lazily tracking the `containous/traefik:experimental` public Docker repository. This is our current running pod:

```
Version:      b0ae6bc049ec4178f012e267ecfd2c72ffd9e7a9
Codename:     roquefort
Go version:   go1.9
Built:        2017-08-29_12:33:38AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

* Kubernetes 1.7
* Traefik deployed via somewhat modified version of the [community Helm chart](https://github.com/kubernetes/charts/tree/master/stable/traefik) (to enable Prometheus metrics and JSON logging)   

Commandline args: `--configfile=/config/traefik.toml`

```toml
# traefik.toml
logLevel = "INFO"
defaultEntryPoints = ["http"]
[entryPoints]
  [entryPoints.http]
  address = ":80"
  compress = false
[kubernetes]
labelselector = "exposure = external"
[accessLog]
format = "json"
[web]
address = ":8080"
[web.metrics.prometheus]
  Buckets=[0.1,0.3,1.2,5.0]
```

### If applicable, please paste the log output in debug mode (`--debug` switch)

Not applicable.

**Comment by timoreimann** [comment]
I think the expectation is missing a _not_.

**Comment by indigoid** [comment]
That's what I would like, yes. I'm looking through the code trying to get my head around how Traefik logging operates. The below in `middlewares/accesslog/logger.go` (`logTheRoundTrip()`) is where the header filtering would need to be done, I think?

```go
for k := range logDataTable.Request {
  fields["request_"+k] = logDataTable.Request.Get(k)
}
```

**Comment by timoreimann** [comment]
/cc @marco.jantke who's quite familiar with the logging logic.

@indigoid in the meantime, could you update the issue description with regards to the _What did you expect to see?_ section please?

**Comment by timoreimann** [comment]
@marco-jantke even ^

**Comment by marco-jantke** [comment]
@indigoid yes this sounds like the right place. Do you think such a filtering should also be done for response headers? Should the header be kept and only the content filtered/replaced for debugging reasons?

A blacklist filter could be made configurable as property of the `types.AccessLog`configuration like this:

```toml
[accessLog]
  format="json"
  headerBlacklist=["X-Auth", "Authorization"]
```

**Comment by indigoid** [comment]
@marco-jantke Yes, filtering of logged response headers is probably also a good idea, though I don't yet have an application for it myself.

What to actually do with the blacklisted headers? I think the most useful action may be to still log the header, so it can be used in log analysis/search, but redact the content.

```json
"request_Authorization": "redacted_by_traefik"
```

**Comment by marco-jantke** [comment]
This sounds good to me 👍 

Do you also want to takeover the implementation for this one? Happy to help along the process and with reviews :)

**Comment by kelmansouri** [comment]
👍

**Comment by indigoid** [comment]
so... after a long hiatus, I was able to find some time to implement this, and have done so. Code works but need to sort out docs etc to match. PR to come. @marco-jantke it will probably need a lot of help as Golang's not my thing

**Comment by ldez** [comment]
Closed in favor of the #2875
